platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java

```diff
@@ -104,19 +112,24 @@ private Optional<WorkspaceResult> loadImmutableWorkspaceIfExists(UnitOfWork work
         }
     }
 
-    private WorkspaceResult loadImmutableWorkspace(UnitOfWork work, File immutableLocation) {
+    private Optional<WorkspaceResult> loadImmutableWorkspaceIfConsistent(UnitOfWork work, ImmutableWorkspace workspace) {
+        File immutableLocation = workspace.getImmutableLocation();
         ImmutableSortedMap<String, FileSystemSnapshot> outputSnapshots = outputSnapshotter.snapshotOutputs(work, immutableLocation);
 
         // Verify output hashes
         ImmutableListMultimap<String, HashCode> outputHashes = calculateOutputHashes(outputSnapshots);
         ImmutableWorkspaceMetadata metadata = workspaceMetadataStore.loadWorkspaceMetadata(immutableLocation);
         if (!metadata.getOutputPropertyHashes().equals(outputHashes)) {
-            throw new IllegalStateException(
-                "Immutable workspace contents have been modified: " + immutableLocation.getAbsolutePath() + ". " +
-                    "These workspace directories are not supposed to be modified once they are created. " +
-                    "Deleting the directory in question can allow the content to be recreated.");
+            boolean moveSuccessful = workspace.withTemporaryWorkspace(temporaryWorkspace -> moveInconsistentImmutableWorkspaceToTemporaryLocation(immutableLocation, temporaryWorkspace, outputSnapshots));
+            if (moveSuccessful) {
```

Comment:
❓ Is it ok we continue if move was not successful or should we fail in that case?

Reply:
Yes, I added a clarifying comment.

Reply:
Thinking about this a bit more, I agree that this is weird. It's also a corner case, and I can't even imagine how it could possibly be triggered. Let's just fall back to failing the build in that case.

Reply:
Clarification: I can only imagine catastrophic ways this could fail, i.e. disk full, permissions on cache directory have changed etc. I can't come up with a scenario we would be expected to recover from.